### PR TITLE
[ENH] Skip unnecessary fit in `ForecastX` if inner `forecaster_y` ignores `X`

### LINKED ISSUE #5040: [ENH] skip training and prediction of `forecaster_X` in `ForecastX` if `forecaster_y` has `ignores-exogeneous-X` as `True`
As of now, `ForecastX` trains and predicts `forecaster_X` always, irrespective of the model specified as `forecaster_y`. But if the endogenous model lacks the capability of using exogenous features, this training and prediction is of no use, and can/should be skipped.

Enhancement Proposal:

1. check tag of `forecaster_y` in `__init__`.
2. if `True`, continue as it is.
3. if `False`, return `forecaster_y` as it is.

Based on my understanding, (3) above should not cause any issue becuase of same API format for all forecasters and pased exogenous features will be ignored.

Optional suggestions:

1. add an argument which will control whether to raise an error, throw a warning, or do nothing.

Documentation issue (unrelated):

1. `forecaster_X` is documented as optional argument which uses `forecaster_y` when not provided explicitly. But this needs to be provided as `None` in that case as it is a positional argument, and can not be skipped in the call altogether.

### PR DESCRIPTION
**Pull request by yarnabrina**
Closes #5040

### CODE REVIEW

**Comment by fkiraly** [comment]
FYI @benHeid, seems related to https://github.com/sktime/sktime/pull/5334.

@yarnabrina, have you checked that there is also no conflict with the current state of #5334?

**Comment by yarnabrina** [comment]
> have you checked that there is also no conflict with the current state of https://github.com/sktime/sktime/pull/5334?

I have not, but it's unlikely. These changes are affecting `forecaster_X`, @benHeid's changes are affecting `forecaster_y`.

File: sktime/forecasting/compose/_pipeline.py [hunk 1]
```diff
@@ -1184,7 +1184,7 @@ class ForecastX(BaseForecaster):
     ----------
     forecaster_y : BaseForecaster
         sktime forecaster to use for endogeneous data `y`
-    forecaster_X : BaseForecaster, optional, default = forecaster_y
```
Comment (fkiraly) [inline sktime/forecasting/compose/_pipeline.py:1187]: oh, I see it was not optional after all.

Should we instead not make the change to `__init__`? It will break nothing downstream.

Out of scope though if you disagree, as you are appropriately correcting a wrong docstring.

File: sktime/forecasting/compose/_pipeline.py [hunk 2]
```diff
@@ -1310,7 +1312,11 @@ def _fit(self, y, X, fh):
         # remember if X seen was None
         self.X_was_None_ = X is None
 
-        if self.behaviour == "update" and X is not None:
+        if (
+            self.behaviour == "update"
+            and X is not None
+            and not self.get_tag("ignores-exogeneous-X")
```
Comment (fkiraly) [inline sktime/forecasting/compose/_pipeline.py:1318]: I believe if the tag is `ignores-exogenous-X`, we already pass `X=None` to both `_fit` and `_predict`, so the third line is superfluous.

File: sktime/forecasting/compose/_pipeline.py [hunk 3]
```diff
@@ -1437,7 +1449,9 @@ def _predict_interval(self, fh, X, coverage):
                 Upper/lower interval end forecasts are equivalent to
                 quantile forecasts at alpha = 0.5 - c/2, 0.5 + c/2 for c in coverage.
         """
-        X = self._get_forecaster_X_prediction(fh=fh, X=X)
+        if not self.get_tag("ignores-exogeneous-X"):
+            X = self._get_forecaster_X_prediction(fh=fh, X=X)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_pipeline.py:1453]: not DRY, instead of these repeated lines, I would move it to the start of in `_get_forecaster_X_prediction`

File: sktime/forecasting/compose/tests/test_pipeline.py [hunk 4]
```diff
@@ -478,10 +478,14 @@ def test_forecastx_logic():
     assert np.allclose(y_pred, y_pred_manual)
 
 
+@pytest.mark.skipif(
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_pipeline.py:481]: can we:

* leave the `NaiveForecaster` as one option, as it ignores `X`
* instead add a test, but use an exogenous forecaster that does not require soft deps, e.g., `YfromX`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks, in-principle ok with this.

Blocking (only small things):

* logic of handling the tag in many predict functions should be more DRY, see above
* changes to tests should not remove case coverage and not rely on soft deps, see above

**Comment by fkiraly** [comment]
small question, I assume this is for a tuning corner case, right? Otherwise, in a manual setting, if `forecaster_y` ignores `X`, you wouldn't use `ForecastX` in the first place, right?

File: sktime/forecasting/compose/_pipeline.py (hunk 1 above)
Comment (yarnabrina) [reply sktime/forecasting/compose/_pipeline.py:1187]: I'll prefer to keep `__init__` as is. I know `sktime` don't use it, but since `python3.8` there are positional-only and keyword-only arguments. I'm not sure if making `forecaster_X` optional now by switching from positional to keyword will break users' workflows or not.

**Comment by yarnabrina** [comment]
> small question, I assume this is for a tuning corner case, right? Otherwise, in a manual setting, if forecaster_y ignores X, you wouldn't use ForecastX in the first place, right?

Yes, mainly for automated tuning case.

File: sktime/forecasting/compose/_pipeline.py (hunk 3 above)
Comment (yarnabrina) [reply sktime/forecasting/compose/_pipeline.py:1453]: Addressed by returning `X` unchanged from `_get_forecaster_X_prediction` in case `ignores-exogeneous-X` is is `True`.

Is that okay, or would you prefer to return `None` instead of unchanged `X`?

Comment (yarnabrina) [reply sktime/forecasting/compose/_pipeline.py:1453]: Realised that this is redundant altogether based on your above comment about `None`. Removing these altogether.

File: sktime/forecasting/compose/_pipeline.py (hunk 2 above)
Comment (yarnabrina) [reply sktime/forecasting/compose/_pipeline.py:1318]: Based on my changes in `_fit`, `self.forecaster_X_` do not exist if `forecaster_y` has `ignores-exogeneous-X` tag as `True`. So, if I don't add this check in `_update`, I thought it'll fail with `AttributeError`.

Now I see it's handled in `_check_X_y` in `BaseForecaster`, which I was not aware of. Will remove from both `_fit` and `_update` in that case, and the change in `_predict_*` (or in `_get_forecaster_X_prediction`) is also redundant in that case in my understanding. Let me know if I understood correctly.

Comment (fkiraly) [reply sktime/forecasting/compose/_pipeline.py:1318]: makes sense. We should e mindful that there are cases where `X` in `predict` and `update` can e none, ut in `fit` it is not (case of no future-known variales)

File: sktime/forecasting/compose/_pipeline.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_pipeline.py:1187]: how could this break sth? I don't get the scenario

File: sktime/forecasting/compose/tests/test_pipeline.py (hunk 4 above)
Comment (yarnabrina) [reply sktime/forecasting/compose/tests/test_pipeline.py:481]: Addressed.

File: sktime/forecasting/compose/tests/test_pipeline.py [hunk 5]
```diff
@@ -499,27 +499,148 @@ def test_forecastx_attrib_broadcast():
     model_1 = model.clone()
     model_1.fit(df[["c"]], X=df[["d", "e"]], fh=[1, 2, 3])
 
-    assert hasattr(model_1, "forecaster_X_")
-    assert isinstance(model_1.forecaster_X_, NaiveForecaster)
-    assert model_1.forecaster_X_.is_fitted
+    assert not hasattr(model_1, "forecaster_X_")
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_pipeline.py:502]: could you make two versions of this, one for the "has forecaster_X_" and one "has not" case?

**Review by fkiraly — CHANGES_REQUESTED** [review]
In the tests, could you kindly make sure that we check both the "ignores `X`" and "doesn't ignore `X`" case now?

File: sktime/forecasting/compose/tests/test_pipeline.py (hunk 5 above)
Comment (yarnabrina) [reply sktime/forecasting/compose/tests/test_pipeline.py:502]: Can you please tell what do you want me to add?

I thought I covered all cases in `test_forecastx_skip_forecaster_X_fitting_logic` and `test_forecastx_flow_known_unknown_columns` (across parametrisation)?

Comment (fkiraly) [reply sktime/forecasting/compose/tests/test_pipeline.py:502]: ah, my mistake. I was confused since `test_forecastx_skip_forecaster_X_fitting_logic` introduces the coverage that you are removing from the first test. All is fine.

**Review by fkiraly — APPROVED** [review]
Blocker was a reviewer error, apologies.
